File: packages/go/hushspec/provider.go
package hushspec

import (
	"context"
	"errors"
	"fmt"
	"os"
	"path/filepath"
	"strings"
	"sync"
	"time"
)

// FileProvider loads a policy from a path on disk, resolving and verifying its
// `extends` chain against the file's own directory.
//
// Resolution happens here rather than in the guard because a file policy's
// relative `extends` references are only meaningful against this file's
// directory, and its detached signature is only findable next to the file
// itself.
type FileProvider struct {
	path    string
	options ResolveOptions
	loader  ResolveLoader
}

// NewFileProvider provides the policy at path, verified under options. The
// zero [ResolveOptions] merges the chain and enforces digest pins without
// looking for a signature.
func NewFileProvider(path string, options ResolveOptions) *FileProvider {
	return &FileProvider{path: path, options: options}
}

// WithLoader resolves `extends` references through loader instead of the
// default (`builtin:` from the embedded rulesets, everything else from the
// filesystem). It returns the provider so it can be chained.
func (p *FileProvider) WithLoader(loader ResolveLoader) *FileProvider {
	p.loader = loader
	return p
}

// Path is the file this provider reads.
func (p *FileProvider) Path() string { return p.path }

// Source is the file this provider reads.
func (p *FileProvider) Source() string { return p.path }

// Load reads, resolves and verifies the policy.
func (p *FileProvider) Load() (*Resolution, error) {
	spec, source, err := loadSpecFile(p.path)
	if err != nil {
		return nil, err
	}
	return ResolveWithOptions(spec, source, p.loader, p.options)
}

// LoadForGuard is [FileProvider.Load], except that a chain which fails the
// signature requirement is returned as a refusal rather than an error, so the
// guard can record every action it denies under it.
func (p *FileProvider) LoadForGuard() (*Resolution, *GuardRefusal, error) {
	return resolveFileForGuard(p.path, p.options, p.loader)
}

// HTTPProvider loads a policy from an HTTPS URL, refetching on every load.
//
// The fetch goes through [NewHTTPLoader], so every rule that loader enforces
// applies here: HTTPS only, the optional host allowlist, the address check
// after DNS with the checked address pinned for the connection, no redirects, a
// size cap and timeouts. ETag revalidation means a poller that ticks every
// minute costs a conditional request, not a full body, while a policy that
// *did* change is still always refetched.
//
// Options carries the trust requirements of the load. A URL is a location and
// never an identity, so a deployment fetching policy over the network should
// set RequireSignature with a keyring: the provider then looks for `<url>.sig`
// through the same transport (signing spec 7.1) and refuses to hand over a
// policy that does not verify.
//
// A remote policy may only extend a `builtin:` ruleset. A remote *base* --
// `extends` naming another URL from a document that itself came over the
// network -- fails closed with a message saying so, rather than being merged
// from a second location the deployment never named. [HTTPProvider.WithLoader]
// allows one.
type HTTPProvider struct {
	url     string
	config  HTTPLoaderConfig
	options ResolveOptions
	fetch   ResolveLoader
	loader  ResolveLoader
}

// NewHTTPProvider provides the policy at url, fetched under config and verified
// under options.
func NewHTTPProvider(url string, config HTTPLoaderConfig, options ResolveOptions) *HTTPProvider {
	if options.SignatureLocator == nil {
		// The policy came over the network, so its sidecar has to as well: the
		// on-disk default would look for a file named after a URL.
		options.SignatureLocator = HTTPSignatureLocator(config)
	}
	return &HTTPProvider{
		url:     url,
		config:  config,
		options: options,
		fetch:   NewHTTPLoader(config),
		loader:  builtinOnlyLoader(),
	}
}

// WithLoader resolves the fetched policy's own `extends` through loader instead
// of the builtin-only default, which is what permits a remote base. It returns
// the provider so it can be chained.
func (p *HTTPProvider) WithLoader(loader ResolveLoader) *HTTPProvider {
	p.loader = loader
	return p
}

// URL is the address this provider fetches from.
func (p *HTTPProvider) URL() string { return p.url }

// Source is the URL the policy came from, and what a receipt's chain and a
// detached-signature lookup name.
func (p *HTTPProvider) Source() string { return p.url }

// Load fetches, resolves and verifies the policy.
func (p *HTTPProvider) Load() (*Resolution, error) {
	loaded, err := p.fetch(p.url, "")
	if err != nil {
		return nil, err
	}
	resolution, err := ResolveWithOptions(loaded.Spec, loaded.Source, p.loader, p.options)
	if err == nil {
		return resolution, nil
	}
	// A base that could not be loaded is the interesting failure here, so name
	// the reference and the document that declared it. A verification failure
	// passes through untouched, so a caller can still read its
	// [SignatureRequiredError].
	var required *SignatureRequiredError
	if loaded.Spec.Extends != nil && !errors.As(err, &required) {
		return nil, fmt.Errorf("failed to resolve 'extends: %s' from %s: %w",
			*loaded.Spec.Extends, p.url, err)
	}
	return nil, err
}

// builtinOnlyLoader serves `builtin:<name>` and refuses everything else, so a
// policy fetched over the network cannot pull a base from a second location.
func builtinOnlyLoader() ResolveLoader {
	return func(reference string, _ string) (*LoadedSpec, error) {
		spec, ok := LoadBuiltin(reference)
		if ok {
			source := reference
			if !strings.HasPrefix(reference, "builtin:") {
				source = "builtin:" + reference
			}
			return &LoadedSpec{Source: source, Spec: spec}, nil
		}
		if strings.HasPrefix(reference, "builtin:") {
			return nil, &NotFoundError{Reference: reference, Message: "unknown builtin ruleset"}
		}
		return nil, &NotFoundError{
			Reference: reference,
			Message: "a policy fetched over the network may only extend a builtin ruleset; " +
				"pass a loader to allow another base",
		}
	}
}

// ---------------------------------------------------------------------------
// Hot reload
// ---------------------------------------------------------------------------

// DefaultWatchInterval is how often a [PolicyWatcher] stats its file.
const DefaultWatchInterval = 2 * time.Second

// DefaultPollInterval is how often a [PolicyPoller] reloads from its provider.
const DefaultPollInterval = 60 * time.Second

// ReloadOptions configures a [PolicyWatcher] or a [PolicyPoller].
type ReloadOptions struct {
	// Interval is the tick period. Zero means [DefaultWatchInterval] for a
	// watcher and [DefaultPollInterval] for a poller.
	Interval time.Duration
	// Guard, when set, has each new policy swapped into it. A swap that fails
	// leaves the previous policy in force and is reported through OnError.
	Guard *Guard
	// OnChange is called with each new resolution, after a successful swap.
	OnChange func(*Resolution)
	// OnError is called with every failure: a file that cannot be read, a
	// chain that will not resolve or verify, a policy that will not compile.
	// The previous policy stays in force.
	OnError func(error)
	// PanicSentinel, when set, is checked on every tick: the kill switch has
	// to work even while the policy source is unreachable, so it is checked
	// before anything is loaded and it fails closed (an unreadable sentinel
	// counts as present).
	PanicSentinel string
}

func (o ReloadOptions) interval(fallback time.Duration) time.Duration {
	if o.Interval > 0 {
		return o.Interval
	}
	return fallback
}

// policyReloader is the machinery a watcher and a poller share: the last good
// policy, the delivery path into a guard, and the ticker goroutine.
type policyReloader struct {
	provider PolicyProvider
	options  ReloadOptions

	// reloadMu serializes a whole reload -- load, swap, record -- so a manual
	// CheckOnce and a tick cannot both decide the same document is new and
	// deliver it twice.
	reloadMu sync.Mutex

	mu      sync.Mutex
	current *Resolution
	// lastHash is the content hash last delivered. It is what makes a reload
	// idempotent: a file rewritten with identical bytes is not a policy change.
	lastHash string

	startMu sync.Mutex
	cancel  context.CancelFunc
	done    chan struct{}
}

func newPolicyReloader(provider PolicyProvider, options ReloadOptions) *policyReloader {
	reloader := &policyReloader{provider: provider, options: options}
	// Seed from the guard: a watcher started against a guard that already
	// holds this policy must not swap the identical document straight back in.
	if options.Guard != nil {
		if resolution := options.Guard.Resolution(); resolution != nil {
			reloader.current = resolution
			reloader.lastHash = resolution.ContentHash
		}
	}
	return reloader
}

// Current is the last policy successfully delivered, or the one the guard
// started with.
func (r *policyReloader) Current() *Resolution {
	r.mu.Lock()
	defer r.mu.Unlock()
	return r.current
}

func (r *policyReloader) report(err error) {
	if err == nil {
		return
	}
	if r.options.OnError != nil {
		r.options.OnError(err)
	}
	if r.options.Guard != nil {
		r.options.Guard.ReportError(err)
	}
}

// reload loads from the provider and delivers the policy when it differs from
// the one in force. It reports whether a new policy took effect.
func (r *policyReloader) reload() (bool, error) {
	r.reloadMu.Lock()
	defer r.reloadMu.Unlock()
	return r.reloadLocked()
}

// reloadLocked is [policyReloader.reload] with reloadMu already held.
//
// Every failure keeps the last good policy: an unreadable, unresolvable,
// unverifiable or uncompilable document must never displace one that worked.
func (r *policyReloader) reloadLocked() (bool, error) {
	resolution, err := r.provider.Load()
	if err != nil {
		err = &PolicyLoadError{Source: r.provider.Source(), Err: err}
		r.report(err)
		return false, err
	}
	if resolution == nil || resolution.Spec == nil {
		err = &PolicyLoadError{
			Source: r.provider.Source(),
			Err:    errors.New("provider returned no policy"),
		}
		r.report(err)
		return false, err
	}

	r.mu.Lock()
	unchanged := resolution.ContentHash != "" && resolution.ContentHash == r.lastHash
	r.mu.Unlock()
	if unchanged {
		return false, nil
	}

	if guard := r.options.Guard; guard != nil {
		if swapErr := guard.SwapPolicy(resolution); swapErr != nil {
			err := &PolicyLoadError{Source: r.provider.Source(), Err: swapErr}
			r.report(err)
			return false, err
		}
	}

	r.mu.Lock()
	r.current = resolution
	r.lastHash = resolution.ContentHash
	r.mu.Unlock()

	if r.options.OnChange != nil {
		r.options.OnChange(resolution)
	}
	return true, nil
}

// checkPanic runs the sentinel check a tick owes the kill switch.
func (r *policyReloader) checkPanic() {
	if r.options.PanicSentinel != "" {
		CheckPanicSentinel(r.options.PanicSentinel)
	}
}

// start runs tick on a ticker until ctx ends or stop is called.
func (r *policyReloader) start(ctx context.Context, interval time.Duration, tick func()) error {
	r.startMu.Lock()
	defer r.startMu.Unlock()
	if r.done != nil {
		return errors.New("policy reload: already started")
	}
	if ctx == nil {
		ctx = context.Background()
	}
	runCtx, cancel := context.WithCancel(ctx)
	done := make(chan struct{})
	r.cancel, r.done = cancel, done

	go func() {
		// Clear the handles on the way out so a reloader whose context was
		// cancelled from outside can be started again; a start that has already
		// replaced them is left alone.
		defer func() {
			close(done)
			r.startMu.Lock()
			if r.done == done {
				r.cancel, r.done = nil, nil
			}
			r.startMu.Unlock()
		}()
		ticker := time.NewTicker(interval)
		defer ticker.Stop()
		for {
			select {
			case <-runCtx.Done():
				return
			case <-ticker.C:
				tick()
			}
		}
	}()
	return nil
}

// stop ends the ticker goroutine and waits for it. It is safe to call when the
// reloader was never started, or has already stopped on its own context.
func (r *policyReloader) stop() {
	r.startMu.Lock()
	cancel, done := r.cancel, r.done
	r.cancel, r.done = nil, nil
	r.startMu.Unlock()
	if cancel == nil {
		return
	}
	cancel()
	<-done
}

// PolicyWatcher reloads a file-backed policy when the file changes.
//
// It polls rather than subscribing to filesystem events: an editor that
// replaces a file by rename, a container that swaps a mounted ConfigMap, and a
// deploy that writes through a symlink all look different to an event API and
// identical to a stat. A tick that sees the same modification time and size
// reads nothing; a tick that sees a change reloads, and a document whose
// content hash is unchanged is not delivered as a swap.
type PolicyWatcher struct {
	*policyReloader
	path string

	lastSize int64
	lastMod  time.Time
	statSeen bool
}

// NewPolicyWatcher watches the file a [FileProvider] reads. Any provider whose
// Source is a filesystem path works; for one that is not, use a
// [PolicyPoller].
func NewPolicyWatcher(provider PolicyProvider, options ReloadOptions) (*PolicyWatcher, error) {
	if provider == nil {
		return nil, errors.New("policy watcher: a provider is required")
	}
	path := provider.Source()
	if path == "" {
		return nil, errors.New("policy watcher: provider has no source path")
	}
	return &PolicyWatcher{
		policyReloader: newPolicyReloader(provider, options),
		path:           filepath.Clean(path),
	}, nil
}

// Path is the file being watched.
func (w *PolicyWatcher) Path() string { return w.path }

// CheckOnce performs exactly one check: the panic sentinel, then the file's
// modification time and size, then a reload if either moved. It reports
// whether a new policy took effect.
//
// It is the whole of what a tick does, exposed so a test can drive reload
// deterministically instead of waiting on a ticker.
func (w *PolicyWatcher) CheckOnce() (bool, error) {
	w.checkPanic()

	// One reload at a time, stat included: two ticks must not both conclude the
	// file is new.
	w.reloadMu.Lock()
	defer w.reloadMu.Unlock()

	info, err := os.Stat(w.path)
	if err != nil {
		err = &PolicyLoadError{Source: w.path, Err: fmt.Errorf("stat: %w", err)}
		w.report(err)
		return false, err
	}

	changed := !w.statSeen || info.Size() != w.lastSize || !info.ModTime().Equal(w.lastMod)
	// The stamp is updated even when the reload below fails: a broken file is
	// reported once, and the next attempt waits for the file to change again.
	w.lastSize, w.lastMod, w.statSeen = info.Size(), info.ModTime(), true

	if !changed {
		return false, nil
	}
	return w.reloadLocked()
}

// Start begins watching on a background goroutine until ctx is cancelled or
// [PolicyWatcher.Stop] is called. It performs one check immediately, so a file
// that changed between the guard's construction and the watch is picked up.
func (w *PolicyWatcher) Start(ctx context.Context) error {
	if err := w.start(ctx, w.options.interval(DefaultWatchInterval), func() {
		_, _ = w.CheckOnce()
	}); err != nil {
		return err
	}
	_, _ = w.CheckOnce()
	return nil
}

// Stop ends the watch and waits for the goroutine to finish. It is safe to
// call more than once.
func (w *PolicyWatcher) Stop() { w.stop() }

// PolicyPoller reloads from a provider on a fixed interval, whatever the
// provider is: an object store, a control-plane API, a file.
//
// A poll that returns a policy with the content hash already in force is not a
// change and is not delivered; a poll that fails leaves the last good policy
// alone.
type PolicyPoller struct {
	*policyReloader
}

// NewPolicyPoller polls provider.
func NewPolicyPoller(provider PolicyProvider, options ReloadOptions) (*PolicyPoller, error) {
	if provider == nil {
		return nil, errors.New("policy poller: a provider is required")
	}
	return &PolicyPoller{policyReloader: newPolicyReloader(provider, options)}, nil
}

// CheckOnce performs exactly one poll: the panic sentinel, then a load. It
// reports whether a new policy took effect.
func (p *PolicyPoller) CheckOnce() (bool, error) {
	p.checkPanic()
	return p.reload()
}

// Start begins polling on a background goroutine until ctx is cancelled or
// [PolicyPoller.Stop] is called. It polls once immediately.
func (p *PolicyPoller) Start(ctx context.Context) error {
	if err := p.start(ctx, p.options.interval(DefaultPollInterval), func() {
		_, _ = p.CheckOnce()
	}); err != nil {
		return err
	}
	_, _ = p.CheckOnce()
	return nil
}

// Stop ends the polling and waits for the goroutine to finish. It is safe to
// call more than once.
func (p *PolicyPoller) Stop() { p.stop() }

File: packages/python/hushspec/provider.py
"""Where a guard's policy comes from, and how it is kept current.

A :class:`PolicyProvider` is the seam between an enforcement point and the
document it enforces: it answers :meth:`~PolicyProvider.load` with a
:class:`~hushspec.resolve.Resolution` -- the resolved policy *and* the evidence
gathered while resolving it -- so a guard built from a provider carries the same
content hash, ``extends`` chain and signature status into every receipt it
writes.

:class:`PolicyWatcher` and :class:`PolicyPoller` keep that policy current.
Both run one tick at a time (:meth:`~_ReloadLoop.check_once`), on a daemon
thread when started, and both fail *safe* rather than open: a document that
cannot be read, parsed, resolved, verified or applied leaves the policy already
in force untouched and is reported through ``on_error``. The last good policy is
never replaced by a bad one, and never by nothing.

The difference between them is only how much work a tick does: the watcher
consults the file's mtime and size first and reloads only when they move, while
the poller reloads on every tick because a provider's source need not be a file
with a mtime at all.
"""

from __future__ import annotations

import threading
from pathlib import Path
from typing import (
    Any,
    Callable,
    Optional,
    Protocol,
    TypeVar,
    Union,
    runtime_checkable,
)

from hushspec.evaluate import check_panic_sentinel
from hushspec.middleware import resolve_policy_resolution
from hushspec.parse import parse_or_raise
from hushspec.resolve import Resolution, ResolveOptions, Resolver

__all__ = [
    "DEFAULT_PANIC_SENTINEL",
    "DEFAULT_WATCH_INTERVAL_S",
    "DEFAULT_POLL_INTERVAL_S",
    "PolicyProvider",
    "FileProvider",
    "CallbackProvider",
    "PolicyWatcher",
    "PolicyPoller",
]

#: The sentinel file ``h2h panic`` creates, and the one a loop consults when
#: asked to check for one without being told where to look.
DEFAULT_PANIC_SENTINEL = ".hushspec_panic"

#: A watcher tick is a ``stat`` unless the file moved, so it can be frequent.
DEFAULT_WATCH_INTERVAL_S = 1.0

#: A poller tick is a full load -- possibly a remote one -- so it is not.
DEFAULT_POLL_INTERVAL_S = 60.0

ChangeHandler = Callable[[Resolution], None]
ErrorHandler = Callable[[Exception], None]


@runtime_checkable
class PolicyProvider(Protocol):
    """Something a policy can be loaded from, with its evidence.

    ``source`` names the origin (a path, a URL, ``memory``) for error messages
    and for the chain a receipt records; :meth:`load` returns the resolved
    policy or raises. A provider that cannot produce a policy must raise rather
    than return a partial or unresolved one -- a guard never holds a document
    whose ``extends`` is still set.
    """

    source: str

    def load(self) -> Resolution:
        """Load, resolve and (as its options demand) verify the policy."""
        ...


class FileProvider:
    """A policy file on disk, resolved against its own directory.

    ``options`` carries the trust requirements of the load -- ``require_signature``
    and the keyring -- so a provider driving hot reload applies the same proof to
    every reloaded document that the first load demanded. Relative ``extends``
    references resolve against the file's directory and a detached signature is
    looked up next to it (``<path>.sig``), exactly as
    :meth:`hushspec.middleware.HushGuard.from_file` does.
    """

    def __init__(
        self,
        path: Union[str, Path],
        options: Optional[ResolveOptions] = None,
        *,
        loader: Optional[Resolver] = None,
        base_dir: Optional[str] = None,
    ) -> None:
        self.path = Path(path)
        #: The absolute path the leaf resolves and verifies from.
        self.source = str(self.path.resolve())
        self.options = options
        self._loader = loader
        self._base_dir = base_dir or str(Path(self.source).parent)

    def load(self) -> Resolution:
        spec = parse_or_raise(self.path.read_text(encoding="utf-8"))
        return resolve_policy_resolution(
            spec,
            loader=self._loader,
            base_dir=self._base_dir,
            source=self.source,
            options=self.options,
        )

    def fingerprint(self) -> Optional[tuple[int, int]]:
        """``(mtime_ns, size)``, or ``None`` when the file cannot be stat'd.

        ``None`` means "cannot tell", so a watcher reloads rather than assuming
        nothing changed: the cheap check may never stand in for the real one.
        """
        try:
            stat = self.path.stat()
        except OSError:
            return None
        return (stat.st_mtime_ns, stat.st_size)

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"FileProvider({self.source!r})"


class CallbackProvider:
    """A provider built from a callable, for sources that are not files.

    The callable returns either a :class:`~hushspec.resolve.Resolution` or the
    policy text, which is parsed and resolved here; anything it raises is
    reported the same way a file read failure would be.
    """

    def __init__(
        self,
        load: Callable[[], Union[Resolution, str]],
        source: str = "memory",
        options: Optional[ResolveOptions] = None,
        *,
        loader: Optional[Resolver] = None,
        base_dir: Optional[str] = None,
    ) -> None:
        self._load = load
        self.source = source
        self.options = options
        self._loader = loader
        self._base_dir = base_dir

    def load(self) -> Resolution:
        loaded = self._load()
        if isinstance(loaded, Resolution):
            return loaded
        spec = parse_or_raise(loaded)
        # `source` stays unset: a document that came from a callable has no file
        # of its own, so a relative `extends` needs an explicit `base_dir` (or a
        # loader) exactly as `HushGuard.from_yaml` requires one.
        return resolve_policy_resolution(
            spec,
            loader=self._loader,
            base_dir=self._base_dir,
            options=self.options,
        )

    def __repr__(self) -> str:  # pragma: no cover - debugging aid
        return f"CallbackProvider({self.source!r})"


#: So ``with PolicyWatcher(...) as watcher`` types as a ``PolicyWatcher``
#: from the one ``__enter__`` both subclasses share.
_LoopT = TypeVar("_LoopT", bound="_ReloadLoop")


class _ReloadLoop:
    """The shared body of the watcher and the poller.

    State advances only on a tick that succeeded end to end: loaded, resolved,
    and accepted by ``on_change``. A tick that fails anywhere leaves the last
    good resolution -- and the fingerprint that would suppress a retry -- exactly
    as they were, so the next tick tries the same document again.
    """

    #: Consult the source's cheap fingerprint before loading.
    _use_fingerprint = False

    def __init__(
        self,
        provider: PolicyProvider,
        interval_s: float = DEFAULT_POLL_INTERVAL_S,
        on_change: Optional[ChangeHandler] = None,
        on_error: Optional[ErrorHandler] = None,
        *,
        panic_sentinel: Optional[Union[str, bool]] = None,
        on_panic: Optional[Callable[[], None]] = None,
    ) -> None:
        if interval_s <= 0:
            raise ValueError(f"interval_s must be positive: {interval_s!r}")
        self._provider = provider
        self._interval_s = interval_s
        self._on_change = on_change
        self._on_error = on_error
        if panic_sentinel is True:
            panic_sentinel = DEFAULT_PANIC_SENTINEL
        self._panic_sentinel: Optional[str] = (
            None if panic_sentinel in (None, False) else str(panic_sentinel)
        )
        self._on_panic = on_panic
        self._panic_reported = False
        self._lock = threading.RLock()
        #: Serializes whole ticks. Reentrant so a tick may be driven from an
        #: ``on_change`` callback without deadlocking; across threads it stops
        #: two ticks from loading at once and the slower one from
        #: overwriting the newer policy.
        self._tick_lock = threading.RLock()
        self._stop = threading.Event()
        self._thread: Optional[threading.Thread] = None
        self._current: Optional[Resolution] = None
        self._fingerprint: Any = None
        #: Ticks that changed the policy, and ticks that failed. Counters make a
        #: background loop observable without a callback per outcome.
        self.reload_count = 0
        self.error_count = 0

    # -- state ------------------------------------------------------------- #

    @property
    def provider(self) -> PolicyProvider:
        return self._provider

    @property
    def source(self) -> str:
        return self._provider.source

    @property
    def interval_s(self) -> float:
        return self._interval_s

    @property
    def current(self) -> Optional[Resolution]:
        """The last resolution that loaded, resolved and applied cleanly."""
        with self._lock:
            return self._current

    @property
    def content_hash(self) -> Optional[str]:
        with self._lock:
            return self._current.content_hash if self._current is not None else None

    @property
    def running(self) -> bool:
        thread = self._thread
        return thread is not None and thread.is_alive()

    def adopt(self, resolution: Resolution) -> None:
        """Seed the loop with a policy loaded elsewhere, without calling back.

        Used when a guard was already built from ``provider.load()``: the loop
        starts from what is in force rather than announcing it as a change.
        """
        with self._lock:
            self._current = resolution
            self._fingerprint = self._stat_fingerprint()

    # -- ticking ------------------------------------------------------------ #

    def check_once(self) -> Optional[Resolution]:
        """Run one tick. Returns the new resolution, or ``None`` for no change.

        Never raises: a failed tick is reported through ``on_error`` and changes
        nothing, because an enforcement point that loses its policy on a bad
        edit is worse than one running a slightly stale good policy.

        Ticks are serialized, so driving one by hand while the loop is running
        cannot interleave two loads and let the slower one win.
        """
        with self._tick_lock:
            return self._tick()

    def _tick(self) -> Optional[Resolution]:
        try:
            self._check_panic()
            fingerprint = self._stat_fingerprint()
            with self._lock:
                unchanged = (
                    self._use_fingerprint
                    and fingerprint is not None
                    and fingerprint == self._fingerprint
                )
            if unchanged:
                return None
            resolution = self._provider.load()
        except Exception as exc:  # noqa: BLE001 - every failure is reportable
            self._report(exc)
            return None

        with self._lock:
            current = self._current
        if current is not None and resolution.content_hash == current.content_hash:
            # The file was touched but says the same thing: remember the new
            # fingerprint so the next tick is a stat again, and announce nothing.
            with self._lock:
                self._fingerprint = fingerprint
            return None

        if self._on_change is not None:
            try:
                self._on_change(resolution)
            except Exception as exc:  # noqa: BLE001
                # The new policy was rejected downstream (it would not compile,
                # or a guard refused it). Keep the last good one and retry.
                self._report(exc)
                return None

        with self._lock:
            self._current = resolution
            self._fingerprint = fingerprint
            self.reload_count += 1
        return resolution

    def _stat_fingerprint(self) -> Any:
        if not self._use_fingerprint:
            return None
        fingerprint = getattr(self._provider, "fingerprint", None)
        return fingerprint() if callable(fingerprint) else None

    def _check_panic(self) -> None:
        """Consult the kill switch before doing anything else with the policy.

        :func:`~hushspec.evaluate.check_panic_sentinel` latches panic mode on and
        fails closed on any stat error it cannot read as "absent".
        """
        if self._panic_sentinel is None:
            return
        active = check_panic_sentinel(self._panic_sentinel)
        with self._lock:
            announce = active and not self._panic_reported
            self._panic_reported = active
        if announce and self._on_panic is not None:
            try:
                self._on_panic()
            except Exception as exc:  # noqa: BLE001
                self._report(exc)

    def _report(self, exc: Exception) -> None:
        with self._lock:
            self.error_count += 1
        if self._on_error is not None:
            try:
                self._on_error(exc)
            except Exception:  # noqa: BLE001 - a broken handler is not fatal
                pass

    # -- lifecycle ---------------------------------------------------------- #

    def start(self, *, load: bool = True) -> Optional[Resolution]:
        """Load once, then tick every ``interval_s`` on a daemon thread.

        The initial load raises: a loop that never had a policy has nothing to
        fall back to, so the caller must hear about it. Every *later* failure is
        reported through ``on_error`` instead. Pass ``load=False`` when the
        policy is already in force (see :meth:`adopt`).
        """
        if self.running:
            raise RuntimeError("already started")
        initial: Optional[Resolution] = None
        if load:
            initial = self._provider.load()
            self.adopt(initial)
        self._stop.clear()
        self._thread = threading.Thread(
            target=self._run,
            name=f"hushspec-{type(self).__name__.lower()}",
            daemon=True,
        )
        self._thread.start()
        return initial

    def stop(self, timeout: Optional[float] = 5.0) -> bool:
        """Stop ticking. Safe to call more than once, and from any thread.

        Returns whether the loop thread is gone. A join that times out keeps
        the thread reference rather than dropping it: forgetting a thread that
        is still inside a tick would make :attr:`running` report ``False`` and
        let a later :meth:`start` run a second loop alongside the first.
        Calling this from inside a tick returns ``False`` -- the loop exits when
        that tick returns, but a thread cannot join itself.
        """
        self._stop.set()
        thread = self._thread
        if thread is None:
            return True
        if thread is threading.current_thread():
            return False
        thread.join(timeout)
        if thread.is_alive():
            return False
        with self._lock:
            if self._thread is thread:
                self._thread = None
        return True

    def _run(self) -> None:
        # wait() returns True only when stop() set the event, so this both
        # sleeps between ticks and wakes immediately on shutdown.
        while not self._stop.wait(self._interval_s):
            try:
                self.check_once()
            except BaseException as exc:  # noqa: BLE001
                # check_once() reports its own failures, so reaching here means
                # something outside a tick's control failed. The loop must
                # survive it: a dead reload thread silently freezes the policy.
                if isinstance(exc, (KeyboardInterrupt, SystemExit)):
                    raise
                self._report(exc)

    def __enter__(self: _LoopT) -> _LoopT:
        self.start()
        return self

    def __exit__(self, *exc_info: Any) -> None:
        self.stop()


class PolicyWatcher(_ReloadLoop):
    """Watches a policy file, reloading it when its bytes change.

    Each tick stats the file first: when mtime and size are unchanged nothing is
    read, parsed or hashed. When they have moved the policy is loaded and its
    content hash compared, so a touch that does not change the document -- or an
    editor's save of identical bytes -- announces nothing.

        >>> guard = HushGuard.from_provider(FileProvider("policy.yaml"), watch=True)

    Or directly, driving anything::

        with PolicyWatcher(FileProvider(path), 0.5, on_change=apply) as watcher:
            ...
    """

    _use_fingerprint = True

    def __init__(
        self,
        provider: PolicyProvider,
        interval_s: float = DEFAULT_WATCH_INTERVAL_S,
        on_change: Optional[ChangeHandler] = None,
        on_error: Optional[ErrorHandler] = None,
        *,
        panic_sentinel: Optional[Union[str, bool]] = None,
        on_panic: Optional[Callable[[], None]] = None,
    ) -> None:
        super().__init__(
            provider,
            interval_s,
            on_change,
            on_error,
            panic_sentinel=panic_sentinel,
            on_panic=on_panic,
        )

class PolicyPoller(_ReloadLoop):
    """Reloads from any provider on a fixed interval.

    Unlike :class:`PolicyWatcher` it makes no assumption that the source is a
    file, so every tick is a full load; only a changed content hash is a change.
    Use it for providers whose source cannot be cheaply fingerprinted.
    """

    _use_fingerprint = False

